src/backend/columnar/columnar_reader.c

```diff
@@ -243,6 +252,105 @@ ColumnarReadNextRow(ColumnarReadState *readState, Datum *columnValues, bool *col
 }
 
 
+/*
+ * ColumnarReadRowByRowNumber reads row with rowNumber from given relation
+ * into columnValues and columnNulls, and returns true. If not such row
+ * exists, then returns false.
+ */
+bool
+ColumnarReadRowByRowNumber(Relation relation, uint64 rowNumber,
+						   Datum *columnValues, bool *columnNulls,
+						   Snapshot snapshot)
+{
+	StripeMetadata *stripeMetadata = FindStripeByRowNumber(relation, rowNumber, snapshot);
+	if (stripeMetadata == NULL)
+	{
+		/* no such row exists */
+		return false;
+	}
+
+	TupleDesc relationTupleDesc = RelationGetDescr(relation);
+	List *relationColumnList = RelationColumnList(relationTupleDesc);
```

Comment:
This seems like it's happening at the wrong place. Shouldn't columnar_tableam.c decide which attributes are needed, and pass those in as a projectedColumnList?

Reply:
fixed by moving the call made to `NeededColumnsList` into `columnar_index_fetch_tuple`